create custom rollkit `ClientState` type

crodriguezvega opened the pull request:
WiP.

crodriguezvega committed:
add client state proto

crodriguezvega committed:
wip: add custom rollkit client state

crodriguezvega committed:
Merge branch 'feat/rollkit' into carlos/17-create-custom-rollkit-clientstate-type

# Conflicts:
#	src/context/mod.rs
#	src/types/mod.rs

crodriguezvega committed:
reorganizing files and bootstrapping traits implementations

rnbguy committed:
TendermintConsensusState over ConsensusState

crodriguezvega committed:
add initialise implementation + fix linter warnings

crodriguezvega committed:
add client type function + cargo fmt

crodriguezvega committed:
implement verify(non)membership

crodriguezvega committed:
implementing verify client message

rnbguy committed:
impl TryFrom and From

rnbguy committed:
use TendermintConsensusStateType

rnbguy committed:
cargo format

crodriguezvega committed:
lint

rnbguy commented:
Hey, cosmos/ibc-rs#1204 relaxes the `TryFrom<Error = ClientError>` restriction. If you are using the latest branch from `ibc-rs`, you can just do this

```rs
use ibc_client_cw::api::ClientType;
use ibc_client_tendermint::consensus_state::ConsensusState as TendermintConsensusState;

use crate::client_state::ClientState as RollkitClientState;

pub struct RollkitClient;

impl<'a> ClientType<'a> for RollkitClient {
    type ClientState = RollkitClientState;
    type ConsensusState = TendermintConsensusState;
}
```

You can refer to [the current implementation of the Tendermint CW client](https://github.com/cosmos/ibc-rs/blob/main/ibc-clients/ics07-tendermint/cw-contract/src/client_type.rs).

PS: We reversed the direction of `Convertible` and dropped `Error` generic. So,

```rs
V::ConsensusStateRef: Convertible<TendermintConsensusStateType, ClientError>,
```
becomes
```rs
TendermintConsensusStateType: Convertible<V::ConsensusStateRef>
```

Sometimes you may need to add

```rs
ClientError: From<<TendermintConsensusStateType as TryFrom<V::ConsensusStateRef>>::Error>
```

for `ClientError` conversion.

crodriguezvega committed:
implementing some more context methods + update to tip of ibc-rs

crodriguezvega committed:
lint

crodriguezvega commented:
> Hey, [cosmos/ibc-rs#1204](https://github.com/cosmos/ibc-rs/pull/1204) relaxes the `TryFrom<Error = ClientError>` restriction. If you are using the latest branch from `ibc-rs`, you can just do this
> 
> ```rust
> use ibc_client_cw::api::ClientType;
> use ibc_client_tendermint::consensus_state::ConsensusState as TendermintConsensusState;
> 
> use crate::client_state::ClientState as RollkitClientState;
> 
> pub struct RollkitClient;
> 
> impl<'a> ClientType<'a> for RollkitClient {
>     type ClientState = RollkitClientState;
>     type ConsensusState = TendermintConsensusState;
> }
> ```
> 
> You can refer to [the current implementation of the Tendermint CW client](https://github.com/cosmos/ibc-rs/blob/main/ibc-clients/ics07-tendermint/cw-contract/src/client_type.rs).
> 
> PS: We reversed the direction of `Convertible` and dropped `Error` generic. So,
> 
> ```rust
> V::ConsensusStateRef: Convertible<TendermintConsensusStateType, ClientError>,
> ```
> 
> becomes
> 
> ```rust
> TendermintConsensusStateType: Convertible<V::ConsensusStateRef>
> ```
> 
> Sometimes you may need to add
> 
> ```rust
> ClientError: From<<TendermintConsensusStateType as TryFrom<V::ConsensusStateRef>>::Error>
> ```
> 
> for `ClientError` conversion.

Thank you for the heads up, @rnbguy. I have [updated to the tip of ibc-rs and implemented the suggested changes](https://github.com/cosmos/rollkit-ibc/pull/22/commits/7a3f18785ad63394492cce2da8d717a6969e3b9f).

crodriguezvega commented on src/client_state/execution.rs:
```diff
@@ -0,0 +1,186 @@
+    let host_timestamp = ExtClientValidationContext::host_timestamp(ctx)?;
+    let host_height = ExtClientValidationContext::host_height(ctx)?;
+
+    let tendermint_consensus_state: TendermintConsensusStateType = consensus_state
+        .clone()
+        .try_into()
+        .map_err(|_: ClientError| ClientError::UnknownConsensusStateType {
+            consensus_state_type: consensus_state.type_url,
+        })?;
```
@rnbguy Would you happen to know if this can be simplified? Doing just `consensus_state.try_into()` was giving me errors about the `ClientError`, and this is what I came up with to work around them...

rnbguy replied:
Very interesting find. Looks like the Rust compiler is getting confused between the implementations for `TryFrom<Any>` and `TryFrom<E::ConsensusStateRef>`. 

For this particular code, you can remove [this](https://github.com/cosmos/rollkit-ibc/pull/22/files/45b9897e1dabe37361f77869eaa11a7dc2c50d75#diff-d773d4d41580ba906c012458fe64a4ef7f9fc874fc594d8e6d3b1774c1475572R87) trait bound, and that allows `.try_into()?`.

If we really needed that trait bound, you could have done:

```rs
let tendermint_consensus_state =
  <TendermintConsensusStateType as TryFrom<Any>>::try_from(consensus_state.clone())?;
```

rnbguy replied:
I pushed a commit for this 31babdb

crodriguezvega replied:
Thanks for the suggestion, @rnbguy! I will try the removal of the trait bound for now. :)

crodriguezvega committed:
implementation of the remaining client state functions

crodriguezvega committed:
lint

crodriguezvega committed:
a bit of cleanup

crodriguezvega commented:
This looks good to me for now. No tests (yet), but we'll figure out how to best do that later on. Merging now...

rnbguy committed:
fix trait bound

rnbguy committed:
use From<ContractError> for StdError

rnbguy committed:
rm unused imports
